kernel/src/hil/gpio.rs

```diff
@@ -176,6 +176,20 @@ pub trait Output {
     }
 }
 
+impl<'a, O: Output> Output for &'a O {
```

Comment:
Well, specifically, `ChipSelectPolar` needs to be able to wrap a type `&'a SOME_SPECIFIC_GPIO` _and_ `ChipSelectPolar` needs to be `Copy`---a constraint imposed by the `type ChipSelect` associated type in `SpiMaster` that's used in the virtual spi stack, though maybe can be worked around.

We could have `ChipSelectPolar<'a, P: gpio::Output>`, but then it won't be `Copy` in cases where the actual type of `P` is not Copy, which is most GPIO implementations (because they typically contain a client which is not Copy). Even if the field is a `&'a P`, the struct still can't be marked Copy, and it won't be usable as a `ChipSelect`.

In other words, the `ChipSelectPolar` needs to name only types that are Copy, which excludes most implementations of `gpio::Output`.

I'm open to suggestions!

Reply:
OK, removed by making `ChipSelectPolar` take a lifetime parameter and always store a reference.

Reply:
@alevy and I discussed this offline, and for posterity's sake, this is an issue because `ChipSelectPolar` wants to take either an owned struct that is `Copy`, or a shared reference (which is `Copy` by definition, despite the compiler's `Copy` derive not determining the trait bounds correctly: https://github.com/rust-lang/rust/issues/26925).

What we'd want from the compiler here is one of the following:
- either allow accepting a reference `&T` where `T: Trait` as a type parameter `P: Trait` (although that would probably open a whole different can of worms),
- or, more conservatively, the ability to accept some `P: Borrow<impl Trait>`. Because owned types generally allow to be borrowed as references to themselves, this trait bound would admit both owned types and references to them.

  Notably, this is already supported, but its ergonomics are quite bad. You can't use `impl Trait` as a type parameter for generic traits and have the compiler plug in the actual type automatically. Instead, you have to write a signature like `T: Trait, P: Borrow<T>`. However, the spurious type parameter `T` will then need to be explicitly specified each time, and also be "used" by introducing a `PhantomData<T>`.

It seems "wrong" to propagate these issues towards the `Output` trait, which `ChipSelectPolar` merely consumes. Notably, if these structs and traits were defined in different crates, an `impl<T: Output> Output for &T { ... }` would not be admitted in the `ChipSelectPolar`-defining crate -- it would require changes to the upstream crate.

Simply always taking a reference works around these issues.